dags/rlc.py

```diff
@@ -0,0 +1,335 @@
+            else:
+                latest_date = datetime_format_dates[0]
+                for date in datetime_format_dates[1:]:
+                    if date > latest_date:
+                        latest_date = date
+                formatted_date = latest_date.strftime("%Y-%m-%d")
+                return formatted_date
+
+def pull_lpi(conn):
+    
+    local_table='vz_safety_programs_staging.signals_cart'
+    url = "https://secure.toronto.ca/opendata/cart/traffic_signals/v3?format=json"
+    return_json = requests.get(url).json()
+    
+    rows = []
+
+    # column names in the PG table
+    col_names = ['asset_type','px','main_street','midblock_route','side1_street','side2_street','latitude','longitude','details','activation_date'] 
+
+    # attribute names in JSON dict
+    att_names = ['px', 'main', 'mid_block', 'side1', 'side2', 'lat', 'long', 'lpiComment']
+
+    # each "info" is all the properties of one LPI, including its coords
+    for obj in return_json:
+        if obj['leading_pedestrian_intervals'] == 1:
+
+            # temporary list of properties of one LPI to be appended into the rows list
+            one_lpi = []
+
+            one_lpi.append('Leading Pedestrian Intervals') # append the asset_name as listed in EC2
+
+            # append the values in the same order as in the table
+            for attr in att_names:
+                one_lpi.append(obj[attr])
+
+            # append the latest LPI implementation date of any of the four directions
+            one_lpi.append(lastest_imp_date(obj))
+
+            rows.append(one_lpi)
+    
+    # delete existing LPI and insert into the local table
+    insert = 'INSERT INTO {} ({}) VALUES %s'.format(local_table, ','.join(col_names))
+    with conn:
+        with conn.cursor() as cur:
+            cur.execute("DELETE FROM {} WHERE asset_type = 'Leading Pedestrian Intervals'".format(local_table))
+            execute_values(cur, insert, rows)
+
+# ------------------------------------------------------------------------------
+# Set up the dag and task
+RLC_DAG = DAG(
+    'rlc_dag',
+    default_args=DEFAULT_ARGS,
+    max_active_runs=1,
+    template_searchpath=[os.path.join(AIRFLOW_ROOT, 'assets/rlc/airflow/tasks')],
+    schedule_interval='0 4 * * 1-5')
+    # minutes past each hour | Hours (0-23) | Days of the month (1-31) | Months (1-12) | Days of the week (0-7, Sunday represented as either/both 0 and 7)
+
+PULL_RLC = PythonOperator(
+    task_id='pull_rlc',
+    python_callable=pull_rlc,
+    dag=RLC_DAG,
+    op_args=[conn]
+)
+
+PULL_APS = PythonOperator(
+    task_id='pull_aps',
+    python_callable=pull_aps,
+    dag=RLC_DAG,
+    op_args=[conn]
+)
+
+PULL_PXO = PythonOperator(
+    task_id='pull_pxo',
+    python_callable=pull_pxo,
+    dag=RLC_DAG,
+    op_args=[conn]
+)
+
+PULL_LPI = PythonOperator(
+    task_id='pull_lpi',
+    python_callable=pull_lpi,
+    dag=RLC_DAG,
+    op_args=[conn]
+)
+
```

Comment:
We need another task to add any new traffic signal (asset type = `Traffic Signals`) to `vz_safety_programs_staging.signals_cart`

Reply:
Should we combine LPI, APS, and traffic signals into one function (aka one task) for efficiency? All three of them are from the same [URL](https://secure.toronto.ca/opendata/cart/traffic_signals/v3?format=json)

We are currently putting them into three separate tasks because it is easier to track errors in case it happens to any one of them. But the same for loop would happen 3 times.

 @radumas